src/_gettsim/groupings.py

```diff
@@ -0,0 +1,155 @@
+
+    p_id_to_fg_id = {}
+    next_fg_id = 0
+
+    for index, current_p_id in enumerate(p_id):
+        # Already assigned a fg
+        if current_p_id in p_id_to_fg_id:
+            continue
+
+        p_id_to_fg_id[current_p_id] = next_fg_id
+
+        current_vg_id = hh_id[index]
+        current_p_id_einstandspartner = p_id_einstandspartner[index]
+        current_p_id_children = p_id_to_p_ids_children.get(current_p_id, [])
+
+        # Assign fg to einstandspartner
+        if current_p_id_einstandspartner >= 0:
+            p_id_to_fg_id[current_p_id_einstandspartner] = next_fg_id
+
+        # Assign fg to children
+        for current_p_id_child in current_p_id_children:
+            child_index = p_id_to_index[current_p_id_child]
+            child_vg_id = hh_id[child_index]
+            child_alter = alter[child_index]
+            child_p_id_children = p_id_to_p_ids_children.get(current_p_id_child, [])
+
+            if (
+                child_vg_id == current_vg_id
+                # TODO (@hmgaudecker): Add correct conditions for grown up children
+                # https://github.com/iza-institute-of-labor-economics/gettsim/pulls/509
+                # TODO(@MImmesberger): Remove hard-coded number
+                # https://github.com/iza-institute-of-labor-economics/gettsim/issues/668
+                and child_alter < 25
+                and len(child_p_id_children) == 0
+            ):
+                p_id_to_fg_id[current_p_id_child] = next_fg_id
+
+        next_fg_id += 1
+
+    # Compute result vector
+    result = [p_id_to_fg_id[current_p_id] for current_p_id in p_id]
+    return numpy.asarray(result)
+
+
+def sn_id_numpy(
+    p_id: numpy.ndarray,
+    p_id_ehepartner: numpy.ndarray,
+    gemeinsam_veranlagt: numpy.ndarray[bool],
+):
+    """
+    Compute a Steuernummer (ID) for each person / couple.
+    """
+    p_id_to_st_id = {}
+    p_id_to_gemeinsam_veranlagt = {}
+    next_st_id = 0
+    result = []
+
+    for index, current_p_id in enumerate(p_id):
+        current_p_id_einstandspartner = p_id_ehepartner[index]
+        current_gemeinsam_veranlagt = gemeinsam_veranlagt[index]
+
+        if (
+            current_p_id_einstandspartner >= 0
+            and current_p_id_einstandspartner in p_id_to_st_id
+        ):
+            gemeinsam_veranlagt_ehepartner = p_id_to_gemeinsam_veranlagt[
+                current_p_id_einstandspartner
+            ]
+
+            if current_gemeinsam_veranlagt != gemeinsam_veranlagt_ehepartner:
+                message = (
+                    f"{current_p_id_einstandspartner} and {current_p_id} are "
+                    "married, but have different values for "
+                    "gemeinsam_veranlagt."
+                )
+                raise ValueError(message)
+
+            if current_gemeinsam_veranlagt:
+                result.append(p_id_to_st_id[current_p_id_einstandspartner])
+                continue
```

Comment:
Last statement in the loop; how can it do anything?

Reply:
The next 4 statements are also part of the loop

Reply:
Sorry for the noise!